### ISSUE TITLE
Redo forces cursor to end of file

### ISSUE THREAD

**Issue by David-Else** [nicoespeon/abracadabra#50]
## Describe the bug

Thanks for awesome extension!

When I do a 'split if statement' it works, I undo it, that works, then redo it and the cursor goes to the bottom of the file.

![out](https://user-images.githubusercontent.com/12832280/69963542-4fbe1380-1508-11ea-98e2-1fd3f67d82af.gif)

- Version of the extension impacted: _v1.2.1_

_🧙‍ Add any other context about the problem here._

**Comment by nicoespeon** [comment]
Thanks for this very nice bug report @David-Else 👍 

I'll investigate what's going on 🧐

**Comment by nicoespeon** [comment]
@all-contributors please add @David-Else for bug

**Comment by David-Else** [comment]
Thanks for adding me as a contributor, I have not seen it done in this way on Github, nice :)

I am working through Refactoring: Improving the Design of Existing Code (2nd Edition), and it is amazing! The more of that that can be automated the better... I can see things getting pretty complicated for the advanced ones, but I dream of a day you can press a button and perform all the complex logic refactorings automatically... that really would be magic!

**Comment by nicoespeon** [comment]
You're welcome, I like how _all-contributors_ acknowledge for all kind of contributions. Raising up relevant bugs helps a lot!

And yes, _Refactoring 2nd Edition_ has a strong influence on the development of this extension. My former experience with IntelliJ refactorings too. As you said, advanced refactorings are not easy, but when there is a recipe it can be automated and save us so much time 😄

**Comment by nicoespeon** [comment]
I gave this one more thoughts. 

TL;DR: 
- that's indeed a problem
- it's hard to solve and I don't think it's a frequent one
- so I won't fix it and I'm closing the issue
- but I'm open to further feedbacks, ideas or PRs to help me solve that

## Detailed explanations

When we do the refactoring, we insert the code, then we mutate the selection to put the cursor at the correct position: https://github.com/nicoespeon/abracadabra/blob/8eb5340a6260130cbf5b44e8ed48c7e1f90e027e/src/editor/adapters/vscode-editor.ts#L43-L46

When we "undo" (Ctrl+Z), then "redo" (Ctrl + ⇧ + Z) the refactoring, it doesn't redo that mutation. So the code is inserted again, but the cursor ends up at the end of the newly inserted code. Because most refactorings would rewrite the whole code (preserving what wasn't modified), the cursor is likely to end up at the end of the file.

I don't know how to make the "redo" understands the selection mutation—if that's even possible. 

The only way I see to kinda-fix this would be to fine tune each refactoring so they only insert code that changed instead of the whole code. But that's much more complex to do it (properly). And even with this, I think the cursor would still end up further than we would expect.

That's why, considering this happens when you "undo", then "redo" a refactoring and knowing the complexity of the fix, I will not fix it unless:
- Many people report that this is a problem for them because they undo-redo refactorings
- Someone come up with a simpler idea to fix this (or a PR)

**Comment by David-Else** [comment]
Thanks for the detailed info!

**Comment by nicoespeon** [comment]
You're welcome. That's only fair to explain why I'm closing this. I don't have a good solution in mind yet and I don't want this to stay open for 2 years without any feedback.

Sorry I couldn't help more for this. I hope I (or someone) will find a simple way to fix that 🤷‍♂

**Comment by David-Else** [comment]
@nicoespeon Hello. I just wanted to add that this issue will hopefully become a non issue with the next update of VS Code, they are adding refactor preview! This is going to be the best update for VS Code refactoring for a long time, so I hope it will all work well with your extension :)

https://github.com/microsoft/vscode/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+Refactor+preview

**Comment by nicoespeon** [comment]
@David-Else that sounds great indeed 👍 

I can see they're working on improving the way refactorings are provided to the users. That's nice. That will probably mean less custom implementation here for more native solutions (e.g. a keybinding to see the list of available refactorings under the cursor position).
